beacon/light/committee_chain.go

```diff
@@ -0,0 +1,528 @@
+		// in between can also be fixed.
+		for p := s.fixedRoots.periods.Next; p < period; p++ {
+			if err := s.fixedRoots.add(batch, p, s.getCommitteeRoot(p)); err != nil {
+				return err
+			}
+		}
+	}
+	if oldRoot != (common.Hash{}) && (oldRoot != root) {
+		// existing old root was different, we have to reorg the chain
+		if err := s.rollback(period); err != nil {
+			return err
+		}
+	}
+	if err := s.fixedRoots.add(batch, period, root); err != nil {
+		return err
+	}
+	if err := batch.Write(); err != nil {
+		log.Error("Error writing batch into chain database", "error", err)
+		return err
+	}
+	return nil
+}
+
+// DeleteFixedRootsFrom deletes fixed roots starting from the given period.
+// It also maintains chain consistency, meaning that it also deletes updates and
+// committees if they are no longer supported by a valid update chain.
+func (s *CommitteeChain) DeleteFixedRootsFrom(period uint64) error {
+	s.chainmu.Lock()
+	defer s.chainmu.Unlock()
+
+	if period >= s.fixedRoots.periods.Next {
+		return nil
+	}
+	batch := s.db.NewBatch()
+	s.fixedRoots.deleteFrom(batch, period)
+	if s.updates.periods.IsEmpty() || period <= s.updates.periods.First {
+		// Note: the first period of the update chain should always be fixed so if
+		// the fixed root at the first update is removed then the entire update chain
+		// and the proven committees have to be removed. Earlier committees in the
+		// remaining fixed root range can stay.
+		s.updates.deleteFrom(batch, period)
+		s.deleteCommitteesFrom(batch, period)
+	} else {
+		// The update chain stays intact, some previously fixed committee roots might
+		// get unfixed but are still proven by the update chain. If there were
+		// committees present after the range proven by updates, those should be
+		// removed if the belonging fixed roots are also removed.
+		fromPeriod := s.updates.periods.Next + 1 // not proven by updates
+		if period > fromPeriod {
+			fromPeriod = period // also not justified by fixed roots
+		}
+		s.deleteCommitteesFrom(batch, fromPeriod)
+	}
+	if err := batch.Write(); err != nil {
+		log.Error("Error writing batch into chain database", "error", err)
+		return err
+	}
+	return nil
+}
+
+// deleteCommitteesFrom deletes committees starting from the given period.
+func (s *CommitteeChain) deleteCommitteesFrom(batch ethdb.Batch, period uint64) {
+	deleted := s.committees.deleteFrom(batch, period)
+	for period := deleted.First; period < deleted.Next; period++ {
+		s.committeeCache.Remove(period)
+	}
+}
+
+// GetCommittee returns the committee at the given period.
+// Note: GetCommittee can be called either with locked or unlocked chain mutex.
+func (s *CommitteeChain) GetCommittee(period uint64) *types.SerializedSyncCommittee {
+	committee, _ := s.committees.get(period)
+	return committee
+}
+
+// AddCommittee adds a committee at the given period if possible.
+func (s *CommitteeChain) AddCommittee(period uint64, committee *types.SerializedSyncCommittee) error {
```

Comment:
Shouldn't this be driven by `InsertUpdate`?

Reply:
Committees can be added in two ways: either together with an update that proves its validity or separately, based on a fixed committee root. This is how the chain is initialized, the bootstrap data proves two committee roots that we can set as fixed, and one full committee that we can add right away via this function. You cannot call `InsertUpdate` until at least one committee has been added.
It can also be used by servers who want to backfill some committees based on historical beacon state proofs that can also prove old committee roots. This is why I exported `AddFixedCommitteeRoot` and `AddCommittee` separately, because root proving can be done in different ways and getting the actual committee might happen at a different time.